ksql-engine/src/test/java/io/confluent/ksql/physical/PhysicalPlanBuilderTest.java

```diff
@@ -739,42 +740,333 @@ public void shouldHaveOptionalFieldsInResultSchema() {
+        + "ON test2.col1 = test3.id;")
+        .get(0);
+
+    // Then:
+    assertThat(result.getExecutionPlan(),
+        containsString("[ REKEY ] | Schema: [TEST2.ROWTIME : BIGINT"));
+  }
+
+  @Test
+  public void shouldRepartitionRightStreamIfNotCorrectKey() {
+    // Given:
+    givenKafkaTopicsExist("test2", "test3");
+    execute(CREATE_STREAM_TEST2 + CREATE_STREAM_TEST3);
+
+    // When:
+    final QueryMetadata result = execute("CREATE STREAM s1 AS "
+        + "SELECT * FROM test2 JOIN test3 WITHIN 1 SECOND "
+        + "ON test2.id = test3.col0;")
+        .get(0);
+
+    // Then:
+    assertThat(result.getExecutionPlan(),
+        containsString("[ REKEY ] | Schema: [TEST3.ROWTIME : BIGINT"));
+  }
+
+  @Test
+  public void shouldThrowIfLeftTableNotJoiningOnTableKey() {
+    // Given:
+    givenKafkaTopicsExist("test4", "test5");
+    execute(CREATE_TABLE_TEST4 + CREATE_TABLE_TEST5);
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage(
+        "Source table (TEST4) key column (TEST4.ID) is not the column "
+            + "used in the join criteria (TEST4.COL0).");
+
+    // When:
+    execute("CREATE TABLE t1 AS "
+        + "SELECT * FROM test4 JOIN test5 "
+        + "ON test4.col0 = test5.id;");
+  }
+
+  @Test
+  public void shouldThrowIfRightTableNotJoiningOnTableKey() {
+    // Given:
+    givenKafkaTopicsExist("test4", "test5");
+    execute(CREATE_TABLE_TEST4 + CREATE_TABLE_TEST5);
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage(
+        "Source table (TEST5) key column (TEST5.ID) is not the column "
+            + "used in the join criteria (TEST5.COL0).");
+
+    // When:
+    execute("CREATE TABLE t1 AS "
+        + "SELECT * FROM test4 JOIN test5 "
+        + "ON test4.id = test5.col0;");
+  }
+
+  @Test
+  public void shouldNotRepartitionEitherStreamsIfJoiningOnKeys() {
+    // Given:
+    givenKafkaTopicsExist("test2", "test3");
+    execute(CREATE_STREAM_TEST2 + CREATE_STREAM_TEST3);
+
+    // When:
+    final QueryMetadata result = execute("CREATE STREAM s1 AS "
+        + "SELECT * FROM test2 JOIN test3 WITHIN 1 SECOND "
+        + "ON test2.id = test3.id;")
+        .get(0);
+
+    // Then:
+    assertThat(result.getExecutionPlan(), not(containsString("[ REKEY ]")));
+  }
+
+  @Test
+  public void shouldNotRepartitionEitherStreamsIfJoiningOnRowKey() {
+    // Given:
+    givenKafkaTopicsExist("test2", "test3");
+    execute(CREATE_STREAM_TEST2 + CREATE_STREAM_TEST3);
+
+    // When:
+    final QueryMetadata result = execute("CREATE STREAM s1 AS "
+        + "SELECT * FROM test2 JOIN test3 WITHIN 1 SECOND "
+        + "ON test2.rowkey = test3.rowkey;")
+        .get(0);
+
+    // Then:
+    assertThat(result.getExecutionPlan(), not(containsString("[ REKEY ]")));
+  }
+
+  @Test
+  public void shouldNotRepartitionEitherStreamsIfJoiningOnRowKeyEvenIfStreamsHaveNoKeyField() {
```

Comment:
just wondering, what happens if you join on ROWKEY and then PARTITION BY ROWKEY? I assume it should work fine, but maybe a good thing to QTT for

Reply:
Yeah, also while working on this I've realised really we need `KeyFields` not `KeyField`, cos there's plenty of scope for to be multiple copies of the key in the projection...

Reply:
Gave it a try and you can't `PARTITION BY ROWKEY` its not valid.